src/schemes/boundary/open_boundary/system.jl

```diff
@@ -0,0 +1,476 @@
+end
+
+function evaluate_characteristics!(system, neighbor_system::FluidSystem,
+                                   v, u, v_ode, u_ode, semi, t)
+    (; volume, sound_speed, characteristics, flow_direction,
+    reference_velocity, reference_pressure, reference_density) = system
+
+    v_neighbor_system = wrap_v(v_ode, neighbor_system, semi)
+    u_neighbor_system = wrap_u(u_ode, neighbor_system, semi)
+
+    nhs = get_neighborhood_search(system, neighbor_system, semi)
+
+    system_coords = current_coordinates(u, system)
+    neighbor_coords = current_coordinates(u_neighbor_system, neighbor_system)
+
+    # Loop over all fluid neighbors within the kernel cutoff
+    for_particle_neighbor(system, neighbor_system, system_coords, neighbor_coords,
+                          nhs) do particle, neighbor, pos_diff, distance
+        neighbor_position = current_coords(u_neighbor_system, neighbor_system, neighbor)
+
+        # Determine current and prescribed quantities
+        rho_b = particle_density(v_neighbor_system, neighbor_system, neighbor)
+        rho_ref = reference_density(neighbor_position, t)
+
+        p_b = particle_pressure(v_neighbor_system, neighbor_system, neighbor)
+        p_ref = reference_pressure(neighbor_position, t)
+
+        v_b = current_velocity(v_neighbor_system, neighbor_system, neighbor)
+        v_neighbor_ref = reference_velocity(neighbor_position, t)
+
+        # Determine characteristic variables
+        density_term = -sound_speed^2 * (rho_b - rho_ref)
+        pressure_term = p_b - p_ref
+        velocity_term = rho_b * sound_speed * (dot(v_b - v_neighbor_ref, flow_direction))
+
+        kernel_ = smoothing_kernel(neighbor_system, distance)
+
+        characteristics[1, particle] += (density_term + pressure_term) * kernel_
+        characteristics[2, particle] += (velocity_term + pressure_term) * kernel_
+        characteristics[3, particle] += (-velocity_term + pressure_term) * kernel_
+
+        volume[particle] += kernel_
+    end
+
+    return system
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::OutFlow)
+    # J3 is prescribed (i.e. determined from the exterior of the domain).
+    # J1 and J2 is transimtted from the domain interior.
+    characteristics[3, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function prescribe_conditions!(characteristics, particle, ::InFlow)
+    # Allow only J3 to propagate upstream to the boundary
+    characteristics[1, particle] = zero(eltype(characteristics))
+    characteristics[2, particle] = zero(eltype(characteristics))
+
+    return characteristics
+end
+
+@inline function update_quantities!(system::OpenBoundarySPHSystem, v, u, t)
+    (; density, pressure, characteristics, flow_direction, sound_speed,
+    reference_velocity, reference_pressure, reference_density) = system
+
+    @threaded for particle in each_moving_particle(system)
+        particle_position = current_coords(u, system, particle)
+
```

Comment:
```suggestion

# Update quantities of boundary particles by evaluating characteristics
```
Something like this?

Reply:
Hm, the characteristics are evaluated prior to `upate_quantities!`.

`# Update quantities as functions of the characteristic variables`
?

Reply:
```
# Update quantities based on the characteristic variables
```
?